src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -84,6 +85,83 @@ public override Task FlushAsync(CancellationToken cancellationToken)
+            }
+
+            // Early check for cancellation
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            // Check out a buffer and start the copy
+            return CopyToAsyncCore(destination, ArrayPool<byte>.Shared.Rent(bufferSize), cancellationToken);
+        }
+
+        private async Task CopyToAsyncCore(Stream destination, byte[] buffer, CancellationToken cancellationToken)
+        {
+            _state.PinReceiveBuffer(buffer);
+            CancellationTokenRegistration ctr = cancellationToken.Register(s => ((WinHttpResponseStream)s).CancelPendingResponseStreamReadOperation(), this);
+            _state.AsyncReadInProgress = true;
+            try
+            {
+                // Loop until there's no more data to be read
+                while (true)
+                {
+                    // Query for data available
+                    lock (_state.Lock)
+                    {
+                        if (!Interop.WinHttp.WinHttpQueryDataAvailable(_requestHandle, IntPtr.Zero))
+                        {
+                            throw new IOException(SR.net_http_io_read, WinHttpException.CreateExceptionUsingLastError());
+                        }
+                    }
+                    int bytesAvailable = await _state.LifecycleAwaitable;
+                    if (bytesAvailable == 0)
+                    {
+                        break;
+                    }
+                    Debug.Assert(bytesAvailable > 0);
+
+                    // Read the available data
+                    cancellationToken.ThrowIfCancellationRequested();
+                    lock (_state.Lock)
+                    {
+                        if (!Interop.WinHttp.WinHttpReadData(_requestHandle, Marshal.UnsafeAddrOfPinnedArrayElement(buffer, 0), (uint)Math.Min(bytesAvailable, buffer.Length), IntPtr.Zero))
+                        {
+                            throw new IOException(SR.net_http_io_read, WinHttpException.CreateExceptionUsingLastError());
+                        }
+                    }
+                    int bytesRead = await _state.LifecycleAwaitable;
+                    if (bytesRead == 0)
+                    {
+                        break;
+                    }
+                    Debug.Assert(bytesRead > 0);
+
+                    // Write that data out to the output stream
+                    await destination.WriteAsync(buffer, 0, bytesRead, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                _state.AsyncReadInProgress = false;
+                ctr.Dispose();
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+
+            // TODO: ReadAsync wasn't/isn't unpinning the buffer.  Is that a bug?
+            // Should it be unpinned there and here?
```

Comment:
@davidsh, ReadAsync wasn't releasing the pinned buffer when the read operation completed.  It seems like that may have been by design, so as to avoid pinning and unpinning the buffer on each read, but as that buffer is user-provided, seems odd.  I kept that behavior, and mirrored it here, but let me know if we should be unpinning in both places.